sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -1247,33 +1315,107 @@ public void startBundle(StartBundleContext c) throws IOException {
+          } else {
+            failures.add(new BigtableWriteException(record, exception));
+          }
+        }
+      };
+    }
+
+    public void retryIndividualRecord(
+        KV<ByteString, Iterable<Mutation>> record, BoundedWindow window) {
+      try {
+        unbatchedWrites.add(
+            bigtableWriter
+                .writeRecordWithoutBatching(record)
+                .whenComplete(
+                    (singleMutationResult, singleException) -> {
+                      if (singleException != null) {
+                        if (singleException instanceof NotFoundException
+                            && !((NotFoundException) singleException).isRetryable()) {
+                          // if we get another NotFoundException, we know this is the bad record.
+                          badRecords.add(
+                              KV.of(new BigtableWriteException(record, singleException), window));
+                        } else {
+                          failures.add(new BigtableWriteException(record, singleException));
+                        }
+                      }
+                    }));
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
     @FinishBundle
     public void finishBundle(FinishBundleContext c) throws Exception {
       try {
+
         if (bigtableWriter != null) {
-          bigtableWriter.close();
+          try {
+            bigtableWriter.close();
+          } catch (IOException e) {
+            // If the writer fails due to a batching exception, but no failures were detected
+            // it means that error handling was enabled, and that errors were detected and routed
+            // to the error queue. Bigtable will successfully write other failures in the batch,
+            // so this exception should be ignored
+            if (!(e.getCause() instanceof BatchingException)) {
+              throw e;
+            }
+          }
           bigtableWriter = null;
         }
 
+        for (CompletionStage<MutateRowResponse> unbatchedWrite : unbatchedWrites) {
+          try {
+            unbatchedWrite.toCompletableFuture().get();
+          } catch (Exception e) {
+            // ignore exceptions here, they are handled by the .whenComplete as part of the
+            // completion stage this .get exists purely to make sure all unbatched writes are
+            // complete.
+          }
+        }
+
+        for (KV<BigtableWriteException, BoundedWindow> badRecord : badRecords) {
+          try {
+            badRecordRouter.route(
+                c,
+                badRecord.getKey().getRecord(),
+                inputCoder,
+                (Exception) badRecord.getKey().getCause(),
+                "Failed to write malformed mutation to Bigtable",
+                badRecord.getValue());
+          } catch (Exception e) {
+            failures.add(badRecord.getKey());
+          }
+        }
+
         checkForFailures();
+
```

Comment:
after finishBundle, shall we empty badRecords and unbatchedWrites to avoid reference leak? (if so then good to check same thing in other DLQ implementations)

Reply:
other implementations don't do their work in finishbundle. I've avoided leaks by having the sets be reset in @startbundle